### PR TITLE
Teach SpanlessEq binding IDs

### LINKED ISSUE #6740: collapsible_match suggestion changes behaviour
Lint name: collapsible_match

Clippy suggests collapsing the nested match in the following code snippet (the full code is [here](https://gitlab.com/glts/spf-milter/-/blob/8cd23bb27df8316edb15bd7e3a3f19d7596044b6/src/config/read.rs#L291-299)):

```rust
pub fn focus_error(error: &ConfigError) -> &dyn Error {
    match error {
        ConfigError::ReadConfig(e) => match e {
            ReadConfigError::Parse(e) => e,
            e => e,
        },
        e => e,
    }
}
```

I expected to see this happen: no Clippy warning

Instead, this happened:

```
warning: Unnecessary nested match
   --> src/config/read.rs:293:39
    |
293 |           ConfigError::ReadConfig(e) => match e {
    |  _______________________________________^
294 | |             ReadConfigError::Parse(e) => e,
295 | |             e => e,
296 | |         },
    | |_________^
    |
    = note: `#[warn(clippy::collapsible_match)]` on by default
help: The outer pattern can be modified to include the inner pattern.
   --> src/config/read.rs:293:33
    |
293 |         ConfigError::ReadConfig(e) => match e {
    |                                 ^ Replace this binding
294 |             ReadConfigError::Parse(e) => e,
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^ with this pattern
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#collapsible_match
```

This suggestion is invalid. It changes the type of the error object is returned and noticeably changes the behaviour of the code (in this case the error message printed to the command-line changes).

### Meta

- `cargo clippy -V`: clippy 0.0.212 (cb75ad5 2021-02-10)
- `rustc -Vv`:
  ```
  rustc 1.50.0 (cb75ad5db 2021-02-10)
  binary: rustc
  commit-hash: cb75ad5db02783e8b0222fee363c5f63f7e2cf5b
  commit-date: 2021-02-10
  host: x86_64-unknown-linux-gnu
  release: 1.50.0
  ```

### PR DESCRIPTION
**Pull request by camsteffen**
changelog: Fix collapsible_match false positive

Fixes #6740

This PR changes the way `SpanlessEq` determines whether two local variables are the same. Instead of checking that the names match, it checks that the `HirId`s match. If local bindings are declared within the expressions that are being compared, `SpanlessEq` will remember bindings that correspond to each other in a `FxHashMap<HirId, HirId>`. This makes `SpanlessEq` more flexible while also fixing false positives. 

Example: `{ let x = 1; x + 2 }` is equal to `{ let y = 1; y + 2 }`.

CC @xFrednet I think this will resolve some concerns in #6463

### CODE REVIEW

File: clippy_lints/src/utils/hir_utils.rs [hunk 1]
```diff
@@ -667,10 +706,15 @@ impl<'a, 'tcx> SpanlessHash<'a, 'tcx> {
...
     }
 
-    pub fn hash_path(&mut self, p: &Path<'_>) {
-        p.is_global().hash(&mut self.s);
-        for p in p.segments {
-            self.hash_name(p.ident.name);
+    pub fn hash_path(&mut self, path: &Path<'_>) {
+        match path.res {
+            // constant hash since equality is dependant on inter-expression context
+            Res::Local(_) => 1_usize.hash(&mut self.s),
```
Comment (flip1995) [inline clippy_lints/src/utils/hir_utils.rs:712]: This is not consistent with the `PartialEq` implementation. `Hash` assumes that all locals are the same, in the `PartialEq` impl they can be different.

Comment (camsteffen) [reply clippy_lints/src/utils/hir_utils.rs:712]: I believe this is ok since a hash function is allowed to have collisions.

Comment (flip1995) [reply clippy_lints/src/utils/hir_utils.rs:712]: Just read the [`Hash` documentation](https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq) and only `Eq` -> `Hash` has to hold up, not `Hash` -> `Eq`. For some reason I thought, that `Hash <-> Eq`. In that case this should be fine.

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Pull request merged**